### PR TITLE
Feature/support conan package creation from local build

### PR DESCRIPTION
**Pull request by maxfaine**
Feature/support conan package creation from local build

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -44,27 +41,42 @@ def requirements(self):
             self.requires("gtest/1.14.0")
             self.requires("boost/1.84.0")
         if self.options.build_unbundled: #each componenet is built independently 
-            self.requires("up-cpp/0.1.1-dev")
-            if self.options.zenoh_package:
-                self.requires("zenohc/cci.20240213")
-        if self.options.build_cross_compiling :
+            self.requires("up-cpp/x.y.z")
             self.requires("zenohc/cci.20240213")
```
Comment (gregmedd) [inline conanfile.py:45]: The change in structure above means that the default is now to include this line, the equivalent of `build_unbundled=True` and `zenoh_package=True` (before, the latter was false). However, zenohc is not available as a package in the conan repositories. The old default behavior was to use up-cpp from the local conan cache and zenoh-c from the system libs (`build_unbundled=True` and `zenoh_package=False`). The only way to get past the `conan install` step now is to set `build_unbundled=False`, but that causes the cmake stage to not work.

Two items need to be addressed:
- The reference to a non-existent conan package should be removed
- README.md should be updated with instructions based on the change in build options such that a developer with a default Ubuntu 22.04 machine could build this library and run the tests.

File: conanfile.py [hunk 2]
```diff
@@ -44,27 +41,42 @@ def requirements(self):
...
 
     def generate(self):
         tc = CMakeToolchain(self)
         tc.variables["BUILD_TESTING"] = self.options.build_testing
-        tc.variables["CROSS_COMPILE"] = self.options.build_cross_compiling
         tc.generate()
 
     def build(self):
+        if os.environ.get("BUILD_DIR_FULL_PATH") is not None:
+            return
```
Comment (gregmedd) [inline conanfile.py:54]: Please provide documentation of the environment and config variables being added to the build system. Any new developer of packager should be able to find out what this does and how to make use of it.

File: conanfile.py [hunk 3]
```diff
@@ -44,27 +41,42 @@ def requirements(self):
...
-        cmake = CMake(self)
-        cmake.install()
+        if os.environ.get("BUILD_DIR_FULL_PATH") is not None:
+            self.output.info("BUILD_DIR_FULL_PATH: " + os.environ.get("BUILD_DIR_FULL_PATH"))
+            build_folder_path = os.environ.get("BUILD_DIR_FULL_PATH")
+            self.output.info("BUILD_DIR_FULL_PATH: " + build_folder_path)
+            self.copy(pattern='*', dst='.', src=('%s/install') % (build_folder_path), symlinks=True, ignore_case=False)
+        else:
+            cmake = CMake(self)
+            cmake.install()
```
Comment (gregmedd) [inline conanfile.py:67]: This is a pretty significant change. Can you provide more of a description of what this does and why it is needed?

File: conanfile.py [hunk 4]
```diff
@@ -44,27 +41,42 @@ def requirements(self):
...
+            self.cpp_info.set_property("cmake_target_name", "up-client-zenoh-cpp::up-client-zenoh-cpp")
+            self.cpp_info.set_property("pkg_config_name", "up-client-zenoh-cpp")
+            self.cpp_info.libs = collect_libs(self)
+
+            self.cpp_info.requires = ["up-cpp::up-cpp", "zenohc::zenohc", "spdlog::spdlog", "protobuf::protobuf"]
+
+            self.cpp_info.names["cmake_find_package"] = "up-client-zenoh-cpp"
+            self.cpp_info.names["cmake_find_package_multi"] = "up-client-zenoh-cpp"
+        else:
+            self.cpp_info.libs = ["up-client-zenoh-cpp"]
```
Comment (gregmedd) [inline conanfile.py:81]: This is a pretty significant change. Can you provide more of a description of what this does and why it is needed?

File: lib/CMakeLists.txt [hunk 5]
```diff
@@ -26,11 +26,7 @@ if(BUILD_UNBUNDLED)
     find_package(up-cpp REQUIRED)
     find_package(zenohc REQUIRED)
 else()
-    if(CROSS_COMPILE)
-        find_package(zenohc REQUIRED)
-    else()
-        find_library(ZENOH_LIB zenohc)
-    endif()
+    find_library(ZENOH_LIB zenohc)
     find_package(up-cpp QUIET)
```
Comment (gregmedd) [inline lib/CMakeLists.txt:30]: This doesn't appear to work with the new changes to conanfile.py. It is now necessary to set `build_unbundled=False` due to the missing zenoh package. This means that up-cpp _will not be a package_ that can be found in this way.

Before the changes to conanfile.py, it was possible to use `build_unbundled=True` and `zenoh_package=False` for `conan install`, then set `BUILD_UNBUNDLED=True` for cmake. That kept up-cpp as a conan dependency, but pulled zenoh-c from the system.

File: test/CMakeLists.txt
Comment (gregmedd) [inline test/CMakeLists.txt:1]: This file has the same build unbundled issue as lib/CMakeLists.txt

**Review by gregmedd — CHANGES_REQUESTED** [review]
This change makes it impossible to build this library on a standard Ubuntu 22.04 machine. I would recommend testing using the instructions in [README.md](/eclipse-uprotocol/up-client-zenoh-cpp/blob/main/README.md) in a clean Ubuntu environment (e.g. `docker run -it ubuntu:22.04`) to verify your changes are good. Feel free to update the README if there are any process changes (e.g. build steps, flags for conan or cmake, etc).

The process of verifying the build steps will start with building and installing zenoh-c.

I'd also like for you to update both the commits and this review with a meaningful description of what these changes are and what they do. The commits can be squashed and reworded however you best feel conveys the content of your work using `git rebase -i ...`.

File: lib/CMakeLists.txt (hunk 5 above)
Comment (MelamudMichael) [reply lib/CMakeLists.txt:30]: @maxfaine  as greg mentioned , this changes were made to support full build

File: conanfile.py [hunk 6]
```diff
@@ -16,55 +16,67 @@ class UpClientZenoh(ConanFile):
...
     def requirements(self):
         self.requires("protobuf/3.21.12" + ("@cross/cross" if self.options.build_cross_compiling else ""))
         self.requires("spdlog/1.13.0")
+        self.requires("up-cpp/x.y.z")
         if self.options.build_testing:
             self.requires("gtest/1.14.0")
             self.requires("boost/1.84.0")
         if self.options.build_unbundled: #each componenet is built independently 
-            self.requires("up-cpp/0.1.1-dev")
-            if self.options.zenoh_package:
```
Comment (MelamudMichael) [inline conanfile.py:48]: this is needed to support full build

File: conanfile.py [hunk 7]
```diff
@@ -16,55 +16,67 @@ class UpClientZenoh(ConanFile):
...
         "build_unbundled": [True, False],
-        "zenoh_package": [True, False],
         "build_cross_compiling": [True, False],
     }
 
     default_options = {
         "shared": False,
         "fPIC": False,
-        "build_testing": True,
+        "build_testing": False,
```
Comment (MelamudMichael) [inline conanfile.py:32]: why this was moved to false?

File: conanfile.py (hunk 3 above)
Comment (maxfaine) [reply conanfile.py:67]: it should enable local package packaging

File: conanfile.py (hunk 7 above)
Comment (maxfaine) [reply conanfile.py:32]: build gtest is optional

File: conanfile.py (hunk 2 above)
Comment (maxfaine) [reply conanfile.py:54]: BUILD_DIR_FULL_PATH enables to wrap local build in conan package

File: conanfile.py (hunk 1 above)
Comment (maxfaine) [reply conanfile.py:45]: build_unbundled=True should be default in general flow

**Comment by gregmedd** [comment]
The CI cache issue with the up-cpp step is fixed by #36. We will get that merged so you can rebase your PR onto the fix.

**Comment by gregmedd** [comment]
Closing stale pull requests. We are tracking conan recipes on the 1.5.8 project board.

**Pull request closed without merging**